src/Npgsql/TypeHandlers/PostgisGeometryHandler.cs

```diff
@@ -0,0 +1,498 @@
+
+                case OgrIdentifier.MultiPolygon:
+                    var mpl = (PostgisMultiPolygon)geom;
+                    if (_ipol == -1)
+                    {
+                        if (_buf.WriteSpaceLeft < 4)
+                            return false;
+                        _buf.PostgisWriteUInt32((uint)mpl.PolygonCount);
+                        _ipol = 0;
+                    }
+                    for (; _ipol < mpl.PolygonCount; _ipol++)
+                    {
+                        if (_irng == -1)
+                        {
+                            if (_buf.WriteSpaceLeft < 9)
+                                return false;
+                            _buf.WriteByte((byte)(BitConverter.IsLittleEndian ? 1 : 0));
+                            _buf.PostgisWriteUInt32((uint)OgrIdentifier.Polygon);
+                            _buf.PostgisWriteUInt32((uint)mpl[_ipol].RingCount);
+                            _irng = 0;
+                        }
+                        for (; _irng < mpl[_ipol].RingCount; _irng++)
+                        {
+                            if (_ipts == -1)
+                            {
+                                if (_buf.WriteSpaceLeft < 4)
+                                    return false;
+                                _buf.PostgisWriteUInt32((uint)mpl[_ipol][_irng].Length);
+                                _ipts = 0;
+                            }
+                            for (; _ipts < mpl[_ipol][_irng].Length; _ipts++)
+                            {
+                                if (_buf.WriteSpaceLeft < 16)
+                                    return false;
+                                _buf.PostgisWriteDouble(mpl[_ipol][_irng][_ipts].X);
+                                _buf.PostgisWriteDouble(mpl[_ipol][_irng][_ipts].Y);
+                            }
+                        }
+                        _irng = -1;
+                    }
+                    return true;
+
+                case OgrIdentifier.GeometryCollection:
+                    var coll = (PostgisGeometryCollection)geom;
+                    if (!_newGeom)
+                    {
+                        if (_buf.WriteSpaceLeft < 4)
+                            return false;
+                        _buf.PostgisWriteUInt32((uint)coll.GeometryCount);
+                        _icol.Push(new Counter());
+                        _newGeom = true;
+                    }
+                    for (Counter i = _icol.Peek(); i < coll.GeometryCount; i.Increment())
+                    {
+                        if (!Write(ref directBuf, coll[i]))
+                            return false;
+                        Reset();
+                    }
+                    _icol.Pop();
+                    return true;
+
+                default:
+                    throw new InvalidOperationException("Unknown Postgis identifier.");
+            }
+        }
+
+        public bool Write(ref DirectBuffer directBuf)
```

Comment:
Any specific reasons for the two Write methods? It seems you can simply reference _toWrite?